docs/src/index.md

```diff
@@ -6,4 +6,356 @@ DocTestSetup = :(using ChainRulesCore, ChainRules)
+or pushforwards
+are not always the same type as the input/outputs of the original function.
+They are differentials,
+differency-equivalents.
+A differential might be such a regular type,
+like a Number, or a Matrix,
+or it might be one of the `AbstractDifferencial` subtypes.
+
+Differentials support a number of operations.
+Most importantly:
+`+` and `*` which lets them act as mathematically objects.
+And `extern` which converts them into a conventional type.
+
+The most important AbstractDifferentials when getting started are the ones about avoiding work:
+
+ - `Thunk`: this is a deferred computation. A thunk is a [word for a zero argument closure](https://en.wikipedia.org/wiki/Thunk). An computation wrapped in a `@thunk` doesn't get evaluated until `extern` is called on the `Thunk`. More on thunks later.
+ - `One`, `Zero`: There are special representions of `1` and `0`. They do great things around avoiding expanding `Thunks` in multiplication and (for `Zero`) addition.
+
+
+
+#### Others: don't worry about them right now
+ - Wirtinger: it is complex. The docs need to be better. [Read the links in this issue](https://github.com/JuliaDiff/ChainRulesCore.jl/issues/40).
+ - Casted: it implements broadcasting mechanics. See [#10](https://github.com/JuliaDiff/ChainRulesCore.jl/issues/10)
+ - InplacableThunk: it is like a Thunk but it can do `store!` and `accumulate!` inplace.
+
+
+ -------------------------------
+## Example of using ChainRules directly.
+
+While ChainRules is largely intended as a backend for Autodiff systems it can be used directly.
+(Infact this can be very useful if you can constraint the code you need to differnetiate to only use thing that have rules defined for.
+This was once how all neural network code worked.)
+
+Using ChainRules directly also helped get a feel for it.
+
+
+```julia
+using ChainRules
+
+function foo(x)
+    a = sin(x)
+    b = 2a
+    c = asin(b)
+    return c
+end;
+
+###
+# Find dfoo/dx via rrules
+
+# First the forward pass, accumulating rules
+x=3;
+a, a_pb = rrule(sin, x);
+b, b_pb = rrule(*, 2, a);
+c, c_pb = rrule(asin, b)
+
+# Then the backward pass calculating gradients
+c̄ = 1;
+_, b̄ = c_pb(extern(c̄));
+_, _, ā = b_pb(extern(b̄));
+_, x̄ = a_pb(extern(ā));
+extern(x̄)
+# -2.0638950738662625
+
+###
+# Find dfoo/dx via frules
+
+# Unlike rrule can interleave evaluation and derivative evaluation
+x=3;
+ẋ=1;
+nofields = NamedTuple();
+
+a, a_pf = frule(sin, x);
+ȧ = a_pf(nofields, extern(ẋ));
+
+b, b_pf = frule(*, 2, a);
+ḃ = b_pf(nofields, 0, extern(ȧ));
+
+c, c_pf = frule(asin, b);
+ċ = c_pf(nofields, extern(ḃ));
+extern(ċ)
+# -2.0638950738662625
+
+###
+# Find dfoo/dx via finite-difference
+using FiniteDifferences
+central_fdm(5,1)(foo, x)
+# -2.0638950738670734
+
+###
+# Via ForwardDiff.jl
+using ForwardDiff
+ForwardDiff.derivative(foo, x)
+# -2.0638950738662625
+
+###
+# Via Zygote
+using Zygote
+Zygote.gradient(foo, x)
+# (-2.0638950738662625,)
+```
```

Comment:
@wesselb @willtebbutt can have Nabla example?
I tried but failed to get Nabla to do this.